src/WLR/Types/Output.hsc

```diff
@@ -0,0 +1,431 @@
+{-# LANGUAGE EmptyDataDeriving, PatternSynonyms #-}
+
+module WLR.Types.Output where
+
+#define WLR_USE_UNSTABLE
+#include <wlr/types/wlr_output.h>
+
+import Foreign
+import Foreign.C.Types
+
+import PIXMAN.Pixman
+import WL.ServerCore
+import WL.ServerProtocol
+import WL.Utils
+import WLR.Backend
+import WLR.Render.Allocator
+import WLR.Render.DrmFormatSet
+import WLR.Render.Interface
+import WLR.Render.Renderer
+import WLR.Render.Swapchain
+import WLR.Render.Texture
+import WLR.Types.Buffer
+import WLR.Types.OutputLayer
+import WLR.Util.Addon
+import WLR.Util.Box
+
+-- A struct from <time.h>. I think this should only be handled by
+-- standard library code, so an opaque pointer should suffice.
+type TIMESPEC = ()
+
+{{ enum
+    WLR_output_mode_aspect_ratio,
+    WLR_OUTPUT_MODE_ASPECT_RATIO_NONE,
+    WLR_OUTPUT_MODE_ASPECT_RATIO_4_3,
+    WLR_OUTPUT_MODE_ASPECT_RATIO_16_9,
+    WLR_OUTPUT_MODE_ASPECT_RATIO_64_27,
+    WLR_OUTPUT_MODE_ASPECT_RATIO_256_135
+}}
+
+{{ struct
+    wlr/types/wlr_output.h,
+    wlr_output_mode,
+    width,                CInt,
+    height,               CInt,
+    refresh,              CInt,
+    preferred,            CBool,
+    picture_aspect_ratio, WLR_output_mode_aspect_ratio,
+    link,                 WL_list
+}}
+
+{{ struct
+    wlr/types/wlr_output.h,
+    wlr_output_cursor,
+    output,      Ptr WLR_output,
+    x,           CDouble,
+    y,           CDouble,
+    enabled,     CBool,
+    visible,     CBool,
+    width,       CUInt,
+    height,      CUInt,
+    src_box,     WLR_fbox,
+    transform,   WL_output_transform,
+    hotspot_x,   CInt,
+    hotspot_y,   CInt,
+    texture,     Ptr WLR_texture,
+    own_texture, CBool,
+    link,        WL_list
+}}
+
+{{ enum
+    WLR_output_adaptive_sync_status,
+    WLR_OUTPUT_ADAPTIVE_SYNC_DISABLED,
+    WLR_OUTPUT_ADAPTIVE_SYNC_ENABLED
+}}
+
+{{ enum
+    WLR_output_state_field,
+    WLR_OUTPUT_STATE_BUFFER,
+    WLR_OUTPUT_STATE_DAMAGE,
+    WLR_OUTPUT_STATE_MODE,
+    WLR_OUTPUT_STATE_ENABLED,
+    WLR_OUTPUT_STATE_SCALE,
+    WLR_OUTPUT_STATE_TRANSFORM,
+    WLR_OUTPUT_STATE_ADAPTIVE_SYNC_ENABLED,
+    WLR_OUTPUT_STATE_GAMMA_LUT,
+    WLR_OUTPUT_STATE_RENDER_FORMAT,
+    WLR_OUTPUT_STATE_SUBPIXEL,
+    WLR_OUTPUT_STATE_LAYERS
+}}
```

Comment:
This is incorrect — the enum names are defined as being equal to powers of two, whereas this defines them as being equal to 0, 1, 2, etc. These should instead be defined similarly to `WLR.Types.Keyboard.WLR_keyboard_led`.

Reply:
I'm not seeing that; do you have that file in a local branch? Regardless, shouldn't it not matter? The macro generates hsc2hs `#const` directives which should get the correct value, no?

Reply:
…yes, you are indeed right about that! `#const` does retrieve the right values.